android/src/main/java/com/learnium/RNDeviceInfo/RNInstallReferrerClient.java

```diff
@@ -0,0 +1,60 @@
+package com.learnium.RNDeviceInfo;
+
+import android.content.SharedPreferences;
+import android.content.Context;
+import android.os.RemoteException;
+
+import com.android.installreferrer.api.InstallReferrerClient;
+import com.android.installreferrer.api.InstallReferrerStateListener;
+import com.android.installreferrer.api.ReferrerDetails;
+
+import static com.android.installreferrer.api.InstallReferrerClient.InstallReferrerResponse;
+
+public class RNInstallReferrerClient implements InstallReferrerStateListener {
+
+  private SharedPreferences sharedPreferences;
+  private InstallReferrerClient referrerClient;
+
+  RNInstallReferrerClient(Context context) {
+    sharedPreferences = context.getSharedPreferences("react-native-device-info", Context.MODE_PRIVATE);
+    referrerClient = InstallReferrerClient.newBuilder(context).build();
+    referrerClient.startConnection(this);
+  }
+
+  private String getInstallReferrer() {
+    try {
+      return referrerClient
+        .getInstallReferrer()
+        .getInstallReferrer();
+    } catch (RemoteException e) {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  referrerClient.startConnection(new InstallReferrerStateListener() {
+    @Override
+    public void onInstallReferrerSetupFinished(int responseCode) {
+      switch (responseCode) {
+        case InstallReferrerResponse.OK:
+          // Connection established
+          SharedPreferences.Editor editor = sharedPreferences.edit();
+          editor.putString("installReferrer", getInstallReferrer());
+          editor.apply();
+          break;
+        case InstallReferrerResponse.FEATURE_NOT_SUPPORTED:
+          // API not available on the current Play Store app
```

Comment:
Everything looks great in this PR, with the one thought that it would be nice to expose the two states (not supported and unavailable) and persist them as well, simply as "not_supported" and "service_unavailable" perhaps?) plus to have a default ("invalid_response" perhaps?) just in case, and document those as possible return values. Seems like we're potentially losing information and may have a higher github issue support load in the future otherwise ("why isn't install referrer returning anything?")